### PR TITLE
Index deletes not applied when cluster UUID has changed

### LINKED ISSUE #11665: Concurrent deletion of indices and master failure can cause indices to be reimported
Currently, a data node deletes indices by evaluating the cluster state. If a new cluster state comes in it is compared to the last known cluster state, and if the new state does not contain an index that the node has in its last cluster state, then this index is deleted.

This could cause data to be deleted if the data folder of all master nodes was lost (#8823): 

All master nodes of a cluster go down at the same time and their data folders cannot be recovered. 
A new master is brought up but it does not have any indices in its cluster state because the data was lost.
Because all other node are data nodes it cannot get the cluster state from them too and therefore sends a cluster state without any indices in it to the data nodes. The data nodes then delete all their data. 

On the master branch we prevent this now by checking if the current cluster state comes from a different master than the previous one and if so, we keep the indices and import them as dangling (see #9952, ClusterChangedEvent).

While this prevents the deletion, it also means that we might in other cases not delete indices although we should.

Example:

1. two masters eligible nodes, m1 is master, one data node (d).
1. m1, m2 and d are on cluster state version 1 that contains and index
1. The index is deleted through the API, causing m1 to send cluster state 2 which does not contain the index to m2 and d that should trigger the actual index deletion.
1. m1 goes down
1. m2 receives the new cluster state but d does not (network issues etc)
1. m2 is elected master and sends cluster state 3 to d which again does not contain the index
1. d will not delete the index because the state comes from a different master than cluster state 1 (the last one it knows of) and will therefore not delete the index and instead import it back into the cluster 

Currently there is no way for a data node to decide if an index should actually be deleted or not if the cluster state that triggers the delete comes from a new master. We chose between: (1) deleting all data in case a node receives an empty cluster state or (2) run the risk to keep indices around that should actually be deleted.

We decided for (2) in #9952. Just opening this issue so that this behavior is documented.

### PR DESCRIPTION
**Pull request by abeyad**
If a node was isolated from the cluster while a delete was happening, the node will ignore the deleted operation when rejoining as we couldn't detect whether the new master genuinely deleted the indices or it is a new fresh "reset" master that was started without the old data folder. We can now be smarter and detect these reset masters and actually delete the indices on the node if its not the case of a reset master.

Note that this new protection doesn't hold if the node was shut down. In that case it's indices will still be imported as dangling indices.

Closes #11665

### CODE REVIEW

**Comment by abeyad** [comment]
@bleskes @dakrone @brwe feedback would be appreciated :)

File: core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java [hunk 1]
```diff
@@ -152,46 +168,35 @@ public boolean indexMetaDataChanged(IndexMetaData current) {
         return true;
     }
 
-    public boolean blocksChanged() {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: Please don't remove this method here, it would make this a breaking change in that it would break any plugins that relied on watching for ClusterChangeEvents where the blocks have changed

File: core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java [hunk 2]
```diff
@@ -152,46 +168,35 @@ public boolean indexMetaDataChanged(IndexMetaData current) {
...
 
+    /**
+     * Returns <code>true</code> iff nodes have been removed from the cluster
+     * since the last cluster state.
+     */
     public boolean nodesRemoved() {
         return nodesDelta.removed();
     }
 
-    public boolean nodesAdded() {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: same here, why did you remove `nodesAdded` and `nodesChanged`?

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: Good point, I only saw they weren't used in ES core so removed them to cleanup, but I'll put them back in (or verify plugins don't use them either, but it is probably not worth doing as part of this commit).

File: core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java [hunk 3]
```diff
@@ -134,10 +142,18 @@ public boolean indexRoutingTableChanged(String index) {
         return deleted == null ? Collections.<String>emptyList() : deleted;
     }
 
+    /**
+     * Returns <code>true</code> iff the metadata for the cluster has changed between
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: can we add a note saying this is an object level equality and not a true equal.

File: core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java [hunk 4]
```diff
@@ -134,10 +142,18 @@ public boolean indexRoutingTableChanged(String index) {
...
+     * the previous cluster state and the new cluster state.
+     */
     public boolean metaDataChanged() {
         return state.metaData() != previousState.metaData();
     }
 
+    /**
+     * Returns <code>true</code> iff the {@link IndexMetaData} for a given index
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: same here re type of equality.

File: core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+        // index routing tables are different because they don't exist
+        newState = createState(numNodesInCluster, randomBoolean(), initialIndices.subList(1, initialIndices.size()));
+        event = new ClusterChangedEvent("_na_", originalState, newState);
+        assertTrue("routing tables should not be the same object", event.routingTableChanged());
+        assertTrue("index routing table should not be the same object", event.indexRoutingTableChanged(initialIndices.get(0)));
+    }
+
+    // Tests that the indices change list is correct as well as metadata equality when the metadata has changed.
+    private static void metaDataChangesCheck(final boolean changeClusterUUID) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: I think we can fold this test into one test and rarely change the cluster uuid?

File: core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        assertTrue("IndexMetaData that didn't previously exist should be considered changed", event.indexMetaDataChanged(newIndexMeta));
+
+        // test when its the same IndexMetaData
+        assertFalse("IndexMetaData should be the same", event.indexMetaDataChanged(originalIndexMeta));
+    }
+
+    /**
+     * Test nodes removed check.
+     */
+    public void testNodesRemoved() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: can we test both node addition and removal (in one cluster state)?

**Comment by bleskes** [comment]
I left some very minor comments. Looks good. I thought we had a test somewhere about index deletion with master change (which we should be able to enable now), but maybe I'm wrong. @brwe do you remember?

**Comment by bleskes** [comment]
also re the pr description:

> This commit fixes the issue by only reimporting indices from data nodes if and only
if the cluster UUID on the master node is different from the cluster
UUID of the previous cluster state on the data node

This inaccurate - on disk dangling indices will still be imported - for example if a node was off while the index was deleted. This change only deals with the case where a node was isolated from the cluster while a delete happened (but was up!) and where a master failure happened during a delete operation (but after the delete was already  committed).   I think we should something like:  

If a node was isloated from the cluster while a delete was happening, the node will ignore the deleted operation when rejoining as we couldn't detect whether the new master genuinely deleted the indices or is a new fresh "resetted" master that was started without the old data folder. We can now be smarter and detect these resetted masters and actually delete the indices if this is not the case. 

Note that this new protection doesn't hold if the node was shut down. In that case it's indices will still be imported as dangling indices.

**Comment by brwe** [comment]
> @brwe do you remember?

I think that is the one that @abeyad enabled here: https://github.com/elastic/elasticsearch/pull/16825/files#diff-c454159f60a0127854028ccb5502500eL1076 I know of no other one.

**Comment by bleskes** [comment]
@brwe I missed that one line in the white space fixing. Jet lag... :( thanks!

File: core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java (hunk 3 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: Done, I also added that note to every other reference equality check in the class

File: core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java (hunk 6 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: Done, this method actually does also check node addition as well, I just didn't name it well because the `ClusterChangedEvent` only contains a `nodesRemoved` method, not a `nodesAdded` method.  So I changed the test method's name.

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: Does it checks the case where nodes are added and removed in one cluster state update? 

Also I'm confused - there is a `org.elasticsearch.cluster.ClusterChangedEvent#nodesAdded` ?

File: core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java (hunk 5 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: @bleskes I'm not clear on what is meant here?  The `testMetaDataChangesOnNoMasterChange` and `testMetaDataChangesOnNewClusterUUID` tests each use this helper method as the test is essentially the same except for whether we change the cluster UUID or not.  I figured we needed a test to check metadata changes when the cluster UUID is changed and one when it is not.

File: core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java (hunk 6 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: @bleskes Sorry you are correct, there is a `nodesAdded`.  What happened was, I originally took it out because I saw it was not used anywhere in core, until @dakrone pointed out that they could be used by plugins and shouldn't be removed.  So I will add a test in here for `nodesAdded` as well (all part of the same test) and include a scenario where nodes are both added and removed in the same cluster state update.

Comment (abeyad) [inline core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java]: Also, I'll include a test for the `nodesChanged` method (as part of the same test).

**Comment by abeyad** [comment]
I fixed the issues raised by @bleskes and @dakrone and all tests pass.  I also updated the PR description and will change the commit message as well once the PR is approved.

**Comment by bleskes** [comment]
LGTM. Thanks @abeyad
